File: global_interp/fortran/polint.f90
  subroutine polint(xa, ya, x, y, dy, ordn)
  implicit none

  integer, intent(in) :: ordn
  real*8, dimension(ordn), intent(in) :: xa, ya
  real*8, intent(in) :: x
  real*8, intent(out) :: y, dy

  integer :: i, m, ns, n_m
  real*8, dimension(ordn) :: c, d, ho
  real*8 :: dif, dift, hp, h, den_val

  c = ya
  d = ya
  ho = xa - x

  ns = 1
  dif = abs(x - xa(1))

  do i = 2, ordn
    dift = abs(x - xa(i))
    if (dift < dif) then
      ns = i
      dif = dift
    end if
  end do

  y = ya(ns)
  ns = ns - 1

  do m = 1, ordn - 1
    n_m = ordn - m
    do i = 1, n_m
      hp = ho(i)
      h  = ho(i+m)
      den_val = hp - h

      if (den_val == 0.0d0) then
        write(*,*) 'failure in polint for point',x
        write(*,*) 'with input points: ',xa
        stop
      end if

      den_val = (c(i+1) - d(i)) / den_val

      d(i) = h * den_val
      c(i) = hp * den_val
    end do

    if (2 * ns < n_m) then
      dy = c(ns + 1)
    else
      dy = d(ns)
      ns = ns - 1
    end if
    y = y + dy
  end do

  return
  end subroutine polint

File: global_interp/c/polint.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

void polint(const double *xa, const double *ya, double x,
            double *y, double *dy, int ordn)
{
    int i, m, n_m;
    int ns_f;
    double dif, dift, hp, h, den_val;

    double *c  = (double*)malloc((size_t)ordn * sizeof(double));
    double *d  = (double*)malloc((size_t)ordn * sizeof(double));
    double *ho = (double*)malloc((size_t)ordn * sizeof(double));
    if (!c || !d || !ho) {
        fprintf(stderr, "polint: malloc failed\n");
        exit(1);
    }

    for (i = 0; i < ordn; i++) {
        c[i]  = ya[i];
        d[i]  = ya[i];
        ho[i] = xa[i] - x;
    }

    ns_f = 1;
    dif = fabs(x - xa[0]);

    for (i = 1; i < ordn; i++) {
        dift = fabs(x - xa[i]);
        if (dift < dif) {
            ns_f = i + 1;
            dif = dift;
        }
    }

    *y = ya[ns_f - 1];
    ns_f -= 1;

    for (m = 1; m <= ordn - 1; m++) {
        n_m = ordn - m;
        for (i = 0; i < n_m; i++) {
            hp = ho[i];
            h = ho[i + m];
            den_val = hp - h;

            if (den_val == 0.0) {
                fprintf(stderr, "failure in polint for point %g\n", x);
                fprintf(stderr, "with input points xa: ");
                for (int t = 0; t < ordn; t++) fprintf(stderr, "%g ", xa[t]);
                fprintf(stderr, "\n");
                exit(1);
            }

            den_val = (c[i + 1] - d[i]) / den_val;
            d[i] = h * den_val;
            c[i] = hp * den_val;
        }

        if (2 * ns_f < n_m) {
            *dy = c[ns_f];
        } else {
            *dy = d[ns_f - 1];
            ns_f -= 1;
        }
        *y += *dy;
    }

    free(c);
    free(d);
    free(ho);
}
